internal/impl/openai/embeddings_processor.go

```diff
@@ -0,0 +1,111 @@
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package openai
+
+import (
+	"context"
+
+	oai "github.com/Azure/azure-sdk-for-go/sdk/ai/azopenai"
+	"github.com/redpanda-data/benthos/v4/public/bloblang"
+	"github.com/redpanda-data/benthos/v4/public/service"
+)
+
+const (
+	oepFieldTextMapping = "text_mapping"
+)
+
+func init() {
+	err := service.RegisterProcessor(
+		"openai_embeddings",
+		embeddingProcessorConfig(),
+		makeEmbeddingsProcessor,
+	)
+	if err != nil {
+		panic(err)
+	}
+}
+
+func embeddingProcessorConfig() *service.ConfigSpec {
+	return service.NewConfigSpec().
+		Categories("AI").
+		Summary("Processor that uses the OpenAI API to create an embedding vector representing the input text.").
+		Description(`
+This processor calls the OpenAI API, computing an embedding of text. By default the entire message's payload as a string is submitted, and the ` + "`text`" + ` configuration field allows customizing that.
+
+You can learn more about embeddings here: https://platform.openai.com/docs/guides/embeddings[https://platform.openai.com/docs/guides/embeddings^]`).
+		Version("4.32.0").
+		Fields(
+			baseConfigFieldsWithModels(
+				"text-embedding-3-large",
+				"text-embedding-3-small",
+				"text-embedding-ada-002",
+			)...,
+		).
+		Field(
+			service.NewBloblangField(oepFieldTextMapping).
+				Description("The text to create embeddings for. By default the entire payload as a string is submitted.").
+				Optional(),
+		)
+}
+
+func makeEmbeddingsProcessor(conf *service.ParsedConfig, mgr *service.Resources) (service.Processor, error) {
+	b, err := makeBaseProcessor(conf)
+	if err != nil {
+		return nil, err
+	}
+	var t *bloblang.Executor
+	if conf.Contains(oepFieldTextMapping) {
+		t, err = conf.FieldBloblang(oepFieldTextMapping)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return &embeddingsProcessor{b, t}, nil
+}
+
+type embeddingsProcessor struct {
+	*baseProcessor
+
+	text *bloblang.Executor
+}
+
+func (p *embeddingsProcessor) Process(ctx context.Context, msg *service.Message) (service.MessageBatch, error) {
+	var body oai.EmbeddingsOptions
+	if p.text != nil {
+		s, err := msg.BloblangQueryValue(p.text)
+		if err != nil {
+			return nil, err
+		}
+		body.Input = append(body.Input, bloblang.ValueToString(s))
+	} else {
+		b, err := msg.AsBytes()
+		if err != nil {
+			return nil, err
+		}
+		body.Input = append(body.Input, string(b))
+	}
+	resp, err := p.client.GetEmbeddings(ctx, body, nil)
+	if err != nil {
+		return nil, err
+	}
+	embd := resp.Data[0]
+	data := make([]any, len(embd.Embedding))
+	for i, f := range embd.Embedding {
+		data[i] = f
```

Comment:
Can you please check if you can access all the members of this data using bloblang? Looking here https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/ai/azopenai#Embeddings, I think `[]EmbeddingItem` and `*EmbeddingsUsage` may be inaccessible via bloblang.

Reply:
Oh, you're right! I think I was looking at something else and then 😵‍💫 My bad! Thanks for checking